### ISSUE TITLE
Tab size setting for diffs

### ISSUE THREAD

**Issue by shaungrady** [desktop/desktop#15561]
### The feature request

Allow users to specify their desired tab size (or tab width) when viewing diffs.

### Proposed solution

In _Preferences_, under _Appearance_, add a **Tab Size** input to allow users to specify their desired tab size, [much like what is offered on GitHub.com's user settings](https://github.com/settings/appearance):

> <img width="707" alt="image" src="https://user-images.githubusercontent.com/52413/199852645-ef33dcf4-af38-40df-8e3c-1589326b474d.png">

### Additional context

This is an important accessibility feature for users who are visually impaired.

**Comment by mettij** [comment]
Hi, im interested in picking this up, can I get assigned this issue?

**Comment by novialriptide** [comment]
<img width="1031" alt="image" src="https://user-images.githubusercontent.com/35881688/206931975-e9516ea9-0fc1-480e-920e-283c82db6e63.png">

I assume it should be placed here?

**Comment by andrewhong04** [comment]
<img width="980" alt="image" src="https://user-images.githubusercontent.com/35881688/206935333-6c6bcdee-4206-4789-88af-c75e4c19a158.png">

I've gotten to work on this, but it looks really out of place.. Is it possible if I could have input from the core developer team? @tidy-dev

**Comment by Daniel-McCarthy** [comment]
An option would be to put it in the drop down menu on the diff page. There's already options regarding whitespace and diff appearance. This menu can be found both under Changes and History when displaying a diff.

![image](https://user-images.githubusercontent.com/4957200/209914134-5b29c2d7-351f-42bc-93e5-4e2ee2cdd68b.png)

**Comment by andrewhong04** [comment]
> An option would be to put it in the drop down menu on the diff page. There's already options regarding whitespace and diff appearance. This menu can be found both under Changes and History when displaying a diff.
> 
> ![image](https://user-images.githubusercontent.com/4957200/209914134-5b29c2d7-351f-42bc-93e5-4e2ee2cdd68b.png)

Thank you! I'll start working on this now.

**Comment by andrewhong04** [comment]
If you guys end up on a decision for the future of this issue, can I be notified? You can also contact me via email on my GitHub profile if you would like to move things forward in private.

**Comment by tidy-dev** [comment]
@novialriptide We got a chance to discuss this issue again and think that what you had suggested before for the Appearance preferences is a good place for it to live.  
![image](https://user-images.githubusercontent.com/75402236/217019482-1ae2f99c-79bf-49e1-88ac-3227e655507f.png)

However, we would like to see headings added of "Theme" above the theme settings and "Diff" above your new tab preference setting. You can check out the Advanced section for example of headings

In addition to that, we would like this to be [feature flagged to beta](https://github.com/desktop/desktop/blob/development/docs/technical/feature-flagging.md). As we think the appearance tab could use a refactor to update the theme section to take up less space and follow other app conventions. Additionally, the new tab preference could be condensed to have a left aligned label.  Thus, we would like to update this before this feature ships to prod (This work is not expected as part of your PR tho, just adding the feature flag).

**Comment by probablykasper** [comment]
Would be nice to support the `TabWidth` `defaults` entry on macOS, so it's possible to have an OS-wide tab width setting (`defaults write -g "TabWidth" 4`)

**Comment by advaith1** [comment]
@novialriptide Any update on this, since the Appearance page has been redesigned?

Also, instead of replacing tabs with spaces like in #15841, it would probably be better to use the [tab-size CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/tab-size).

**Comment by medallyon** [comment]
bumping to get an update on this issue

**Comment by LukeWCS** [comment]
Is there any progress with the decision here? It would be very useful if the tab width could be freely defined so that I can set it to the project specifications, which is not 8 spaces. Support for EditorConfig would be even better, because then you could set it separately for each repo and would therefore be project-specific.

**Comment by ahmetsait** [comment]
It seems that the dev team is more concerned with adding pretty check marks and moving staging buttons around.

**Comment by shaungrady** [comment]
> Support for EditorConfig would be even better, because then you could set it separately for each repo and would therefore be project-specific.

Maybe support for EditorConfig could be an option, but it really should be in the hands of the user to specify the spacing they prefer to support the visually impaired.

**Comment by LukeWCS** [comment]
> Maybe support for EditorConfig could be an option, but it really should be in the hands of the user to specify the spacing they prefer to support the visually impaired.

But not in projects where this is defined as a fixed project specification in order to get easily interchangeable code that looks the same for everyone. As developers, we have strict guidelines that we have to adhere to and with EditorConfig we can easily take this into account project-wide. For example, it is also strictly specified which type of indentation (tabs or spaces) must be used for which file format and also the exact width. Also which line ending must be used and other things. If these guidelines are not followed, code is either rejected or adapted accordingly so that the code complies with the guidelines.

In other words: I don't make these guidelines, I just have to follow them. ^^ That's why it would be immensely helpful if GH Desktop offered this in the same way that GitHub already offers itself.

In this case too, you could freely choose whether to use the project-specific `.editorconfig` file in the repo or not. But at the moment you don't have that choice because EditorConfig is not supported.

**Comment by shaungrady** [comment]
> But not in projects where this is defined as a fixed project specification in order to get easily interchangeable code that looks the same for everyone.

The tab-width setting is purely visual and completely personal—it has no functional impact on the project or the code. People who are visually impaired may require tab-widths of 4, 8, 16 or greater to help them read the code, whereas others may wish to set tabs to be displayed as the equivalent of 2 spaces. This is why this feature is important.

`.editorconfig` is for defining formatting consistency of the code, not for defining the visual representation of code within the editor.

(Defining a tab-width does not apply when spaces are used as indentation, so it has no bearing on this feature.)

**Comment by LukeWCS** [comment]
> The tab-width setting is purely visual and completely personal—it has no functional impact on the project or the code.

I know that. ;) But it has nothing to do with project-wide specifications.

> People who are visually impaired may require tab-widths of 4, 8, 16 or greater to help them read the code, whereas others may wish to set tabs to be displayed as the equivalent of 2 spaces. This is why this feature is important.

But that only applies to the first indentation level. As soon as you have more than one indentation level in a line, the width of an indentation is relevant. That's why it's necessary to be able to define the standard width of an indentation project-wide, because otherwise, depending on the individual settings, you'll see code that looks "frayed" from the second indentation level onwards, and that makes reading it much more difficult depending on the code.

> `.editorconfig` is for defining formatting consistency of the code, not for defining the visual representation of code within the editor.
> 

That's not correct. EditorConfig is also used for the visual representation of the code and that's one of the main reasons why I use EditorConfig. Otherwise, EditorConfig wouldn't have the `indent_size` property. That works exactly as it should with GH Web, but unfortunately not with GH Desktop, which makes viewing diffs unnecessarily difficult. And that works exactly the same in all programs that support EditorConfig; with Notepad++, for example, EditorConfig always has priority.

But okay, your last comment has now made it clear to me that your wish and mine is not an OR question, but an AND question. ^^ I will create a separate issue, since these are two independent properties.

**Comment by dxkf177** [comment]
，
